src/lithium/reducer.py

```diff
@@ -365,44 +377,65 @@ def run(self, testcase, interesting, tempFilename):  # pylint: disable=invalid-n
-                elif empty or last:
-                    log.info("Lithium result: succeeded, reduced to: %s", quantity(len(testcase.parts), testcase.atom))
+
+                # If the testcase is empty, end minimization
+                if testcase.length == 0:
+                    log.info("Lithium result: succeeded, reduced to: %s", quantity(testcase.length, testcase.atom))
                     break
+
+                # If the chunk_size equals or exceeds the minimum and
+                if chunk_size <= min_chunk_size:
+                    # Repeat mode is last or always and atleast one chunk was removed during the last round, repeat
+                    if removed_chunks and (self.minimizeRepeat == "always" or self.minimizeRepeat == "last"):
+                        log.info("Starting another round of chunk size %d", chunk_size)
+                        chunk_end = testcase.length
+                    # Otherwise, end minimization
+                    else:
+                        log.info("Lithium result: succeeded, reduced to: %s", quantity(testcase.length, testcase.atom))
+                        break
+                # If none of the conditions apply, reduce the chunk_size and continue
                 else:
-                    chunkStart = 0
-                    while chunkSize > 1:  # smallest valid chunk size is 1
-                        chunkSize >>= 1
+                    chunk_end = testcase.length
+                    while chunk_size > 1:  # smallest valid chunk size is 1
+                        chunk_size >>= 1
                         # To avoid testing with an empty testcase (wasting cycles) only break when
                         # chunkSize is less than the number of testcase parts available.
-                        if chunkSize < len(testcase.parts):
+                        if chunk_size < testcase.length:
                             break
-                    log.info("Reducing chunk size to %d", chunkSize)
-                anyChunksRemoved = False
-
-            chunkEnd = min(len(testcase.parts), chunkStart + chunkSize)
-            description = "Removing a chunk of size %d starting at %d of %d" % (
-                chunkSize, chunkStart, len(testcase.parts))
-            testcaseSuggestion = testcase.copy()
-            testcaseSuggestion.parts = testcaseSuggestion.parts[:chunkStart] + testcaseSuggestion.parts[chunkEnd:]
-            if interesting(testcaseSuggestion):
-                testcase = testcaseSuggestion
-                log.info("%s was a successful reduction :)", description)
-                anyChunksRemoved = True
-                # leave chunkStart the same
+
+                    log.info("Reducing chunk size to %d", chunk_size)
+                removed_chunks = False
+
+                # Perform post round clean-up if defined
+                temp_testcase = testcase.copy()
+                if self.post_round_cb(temp_testcase):
```

Comment:
This is fine for now, but I'm not sold that this is a good way to do this. Other `Minimise` sub-classes just re-implement the reduce loop which won't call this method (or use your other new loop changes). I think we should abstract out the minimize loop so that all `Minimise` implementors can put their special sauce in without having to copy the whole loop with all it's nuance and hooks like this.

Reply:
Absolutely agree.  This is just a band-aid until I can refactor the class and subsequent strategies.